### ISSUE TITLE
Runtime type for an array of arrays

### ISSUE THREAD

**Issue by vasslitvinov** [chapel-lang/chapel#11220]
This issue is to discuss the solution to #8152 and #11039
that is currently being implemented, starting with #10941.

#### Background: what's in a "runtime type"?

A value of a runtime type contains what's needed to create,
at runtime, a domain or an array. Roughly:

* RTT(D:domain) = D.distribution
* RTT(A:[]) = (A.domain, RTT(A.eltType))
* RTT(other) = nothing

#### The context

Since #8073, a for-/forall-/promoted expression can be used
to create an array of arrays, for example:

```chpl
proc makeArray(i:int) { return (some array); }
var AofA = for i in (whatever) do makeArray(i);
```

This scenario differs from other arrays of arrays because the type
of the resulting `AofA` is not declared by the user.

The object of this design issue is **the domain to use
for the runtime type of `AofA.eltType`**, given that
`AofA.eltType` is itself an array.

Cf. the domain of `AofA` itself is not discussed here.
For now we are OK with how it is computed.

At compile time, `AofA.eltType` is clearly `makeArray(i).type`.
At run time, however, each call to makeArray() may return
an array with its own domain. BTW this may happen even when
the arrays from all makeArray() calls have the same shape,
for example:

```chpl
    proc makeArray(i:int) {
      var A: [1..0] int; // this creates a fresh domain {1..0},
			 // which is contained in A's runtime type
      return A;
    }
```

So, which of these domains to use for AofA.eltType's runtime type?
None of them?

#### Prospective solution (A)

Use the domain of the array from the **first** call to makeArray().

This is what we currently do in the shape-less case.

Pros:
* It comes from an existing array.
* It may be the domain of **all** arrays produced by makeArray(),
  depending on what makeArray() does internally.

Cons:
* Undefined when for loop creating `AofA` executes 0 iterations.
* Poorly defined when that loop is parallel.
* Difficult or costly to implement when we want to create `AofA` **before**
  computing its elements. For example, when `AoA` is block-distributed
  and is initialized from a forall-loop.

#### Prospective solution (B)

Create an empty domain of an appropriate static type.
For example, for `domain(2)`, it is `{1..0,1..0}`.

This is what we currently do in the shape-ful case.

Pros:
* Uniform, does not depend on what elements AofA.
* Works well with the current implementation, where `AofA` is created
  before computing its elements.

Cons:
* Needs to be implemented separately for each domain map.
* Currently this is implemented only for Default Rectangular.
* Current implementation leaks such a domain in many cases.

#### Prospective solution (C)

We could make an array-of-arrays be a different kind of thing altogether
when its type is not declared by the user. For example, allow plurality
of runtime types, one for each element. This needs more design work.

#### Bottom line

The current proposal is to proceed with (B), introduced #10941.
Once loops over ranges are moved to the shapeful world,
this solution will apply in the majority of cases where
an array-of-arrays is created with a loop expression.

**Comment by bradcray** [comment]
My intuition was C:  That the runtime type of `[i in 1..n] [1..i] real` should be "an array over {1..n} of arrays whose domains are {1..1}, {1..2}, {1..3}, ..., {1..n} of reals.

**Comment by bradcray** [comment]
[For those getting email notifications, I realized a mistake in my original note above that the runtime type shouldn't be an array of arrays but rather an array of runtime types.  Attempted to reword to clarify this]

**Comment by mppf** [comment]
Run-time types for element types for arrays of arrays seem to flow through OK in the current compiler:

``` chapel
var A:[1..3] [1..2] int;
  
var B:A.type;

var B1:B[1].type;

writeln(B1);
writeln(B1.domain);
writeln(B1.eltType:string);

var B2:B[2].type;

writeln(B1);
writeln(B1.domain);
writeln(B1.eltType:string);
```

Regarding skyline arrays, I wonder if we should add an idea of arrays that don't have a particular run-time type (or, put another way, the domain is not considered part of the array's "type" and is rather solely owned by that inner array).

That would enable the language / compiler to distinguish between array-of-arrays where the inner array type should be fixed and the case where it should be allowed to vary.

Here's a straw-man syntax proposal (that I know has problems):

``` chapel
var A:[1..3] [D] int; // inner array shape is shared by all inner arrays
var A:[1..3] [?] int; // inner array shape can vary
```

Another way to put this is that in an array declaration, the domain serves two different purposes:
 1. Serving as the initial value for the array's domain
 2. Expressing a run-time link between the array and its domain; that is, requiring the array to have the same shape as any other arrays over that domain.

We could seek to allow one to express (1) without also expressing (2).

How does that help with skyline arrays? Because when the compiler sees
``` chapel
var A = [i in 1..10] makeArray(j);
```

it could infer the type of A as
``` chapel
var A:[1..10] [?] int
```

rather than as
``` chapel
var A:[1..10] [D] int
```

for some D.

What about skyline arrays declared using the so-far-unsupported syntax Brad described?
``` chapel
var A: [i in 1..n] [1..i] int;
```

For such declarations, it should be clear to the compiler that there's not a single domain for the inner arrays. The range 1..i can be used to initialize each inner array's domain.

Actually maybe all of this comment is off-topic since even for my "inner arrays have varying domains" case, we could seek to make a declaration like `var B: A.type;` copy the inner array sizes. Even so I think what I propose in this comment is somewhere between B and C in the original issue description, since I'm bringing up the possibility that maybe the domain should not be considered part of the array type for certain arrays.

**Comment by vasslitvinov** [comment]
Even if we create an array without creating its runtime type first, this question remains: what should be `A.eltType` ?

**Comment by mppf** [comment]
> that are worry about the runtime type may be misguided. Specifically, maybe we try to create it only so that we can create an array out of it.

I think this is reasonable. The implementation currently uses runtime types to create arrays-of-arrays / skyline arrays, but it doesn't have to work that way. Runtime-types are only really necessary for a case like `var B: OtherArray.type`. It would probably be reasonable to leave that pattern not working yet, while getting the compiler-generated initialization patterns to function correctly and without leaks.

> Even if we create an array without creating its runtime type first, this question remains: what should be A.eltType ?

Can you elaborate on why this is still an important question?

**Comment by vasslitvinov** [comment]
> > Even if we create an array without creating its runtime type first, this question remains: what should be A.eltType ?
> 
> Can you elaborate on why this is still an important question?

Because:

* What if somebody says `var B: A.eltType;` - what should `B` be?

* Or a more sophisticated version where `A.eltType` is returned out from a `type` function + passed around, then `B` is declared to be of that type.

* The DSI Array class for `A` contains a field `eltType` storing a `_RuntimeType`. What do we put in it? That `_RuntimeType` instance is used to create `B` in `var B: A.eltType;` or the more sophisticated scenario above.

**Comment by vasslitvinov** [comment]
I agree with @mppf suggestion above that we may choose not to support `var B: A.type;` or `var B: A.eltType;` for now. This would allow us to postpone this design question until there is a demand for this feature. Meanwhile the compiler would refuse to create `B`.

As a workaround, the user would need to declare the type of `A` **explicitly** in order to create a variable of `A.type` or `A.eltType`.

BTW in this setting availability of`A.type` is equivalent to availability of `A.eltType`. That is because the only difference between the two is `A.domain`, which IS available.

**Comment by vasslitvinov** [comment]
Discussing this with Brad and Michael today, we agreed to keep this an open design issue with a lower priority. 

In the short term, report an error whenever a problematic runtime type is accessed. If possible, report the error at compile time, otherwise at run time when bounds checking is enabled. Implementation is T.B.D. We accept this short-term solution in order to allow progress with shape preservation #10899 etc.

**Comment by vasslitvinov** [comment]
P.S. Creating arrays of arrays using for/forall-expressions -- the functionality that helped create this issue -- has been available since #8073.
